File: linux/core/info/internet_info.cc
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <unistd.h>

using namespace std;

class NetworkMonitor
{
public:
    NetworkMonitor()
    {
        // Initialize initial values
        readStats(m_InitialUploadBytes, m_InitialDownloadBytes);
    }

    pair<double, double> getInternetSpeed()
    {
        // Get current values
        unsigned long long currentUploadBytes, currentDownloadBytes;
        readStats(currentUploadBytes, currentDownloadBytes);

        // Calculate upload and download speeds
        double uploadSpeed = (currentUploadBytes - m_InitialUploadBytes) / 1024.0;       // Convert to KB
        double downloadSpeed = (currentDownloadBytes - m_InitialDownloadBytes) / 1024.0; // Convert to KB

        // Update initial values for next calculation
        m_InitialUploadBytes = currentUploadBytes;
        m_InitialDownloadBytes = currentDownloadBytes;

        return make_pair(uploadSpeed, downloadSpeed);
    }

private:
    unsigned long long m_InitialUploadBytes;
    unsigned long long m_InitialDownloadBytes;

    void readStats(unsigned long long &uploadBytes, unsigned long long &downloadBytes)
    {
        ifstream file("/proc/net/dev");
        if (!file.is_open())
        {
            cerr << "Failed to open /proc/net/dev" << endl;
            return;
        }

        // Skip the first two lines (headers)
        string line;
        getline(file, line);
        getline(file, line);

        // Extract interface name and stats
        while (getline(file, line))
        {
            istringstream iss(line);
            string interface;
            iss >> interface;

            if (interface.find(':') != string::npos)
            {
                unsigned long long rxBytes, txBytes;
                iss >> rxBytes >> txBytes;

                uploadBytes += txBytes;
                downloadBytes += rxBytes;
            }
        }

        file.close();
    }
};

int main()
{
    NetworkMonitor monitor;

    // Update speeds every second
    while (true)
    {
        auto speeds = monitor.getInternetSpeed();
        cout << "Upload Speed: " << speeds.first << " KB/s" << endl;
        cout << "Download Speed: " << speeds.second << " KB/s" << endl;
        sleep(1);        // Sleep for 1 second
        system("clear"); // Clear the console
    }

    return 0;
}

File: lib/core/internet_info.dart
import 'dart:async';
import 'dart:io';

class NetworkMonitor {
  late BigInt _initialUploadBytes;
  late BigInt _initialDownloadBytes;

  NetworkMonitor() {
    // Initialize initial values
    _initialUploadBytes = _readStats()[0];
    _initialDownloadBytes = _readStats()[1];
  }

  Future<List<double>> getInternetSpeed() async {
    // Get current values
    var currentStats = _readStats();
    BigInt currentUploadBytes = currentStats[0];
    BigInt currentDownloadBytes = currentStats[1];

    // Calculate upload and download speeds
    double uploadSpeed = (currentUploadBytes - _initialUploadBytes).toDouble() /
        1024.0; // Convert to KB
    double downloadSpeed =
        (currentDownloadBytes - _initialDownloadBytes).toDouble() /
            1024.0; // Convert to KB

    // Update initial values for next calculation
    _initialUploadBytes = currentUploadBytes;
    _initialDownloadBytes = currentDownloadBytes;

    return [uploadSpeed, downloadSpeed];
  }

  List<BigInt> _readStats() {
    BigInt uploadBytes = BigInt.zero;
    BigInt downloadBytes = BigInt.zero;

    var lines = File('/proc/net/dev').readAsLinesSync();
    // Skip the first two lines (headers)
    lines.removeRange(0, 1);

    // Extract interface name and stats
    for (var line in lines) {
      var parts = line.trim().split(RegExp(r'\s+'));
      if (parts.isNotEmpty) {
        String interface = parts[0].replaceAll(':', '');

        BigInt rxBytes = BigInt.tryParse(parts[1]) ?? BigInt.zero;
        BigInt txBytes = BigInt.tryParse(parts[9]) ?? BigInt.zero;

        uploadBytes += txBytes;
        downloadBytes += rxBytes;
      }
    }

    return [uploadBytes, downloadBytes];
  }
}

void main() {
  NetworkMonitor monitor = NetworkMonitor();

  // Update speeds every second
  Timer.periodic(Duration(seconds: 1), (timer) async {
    for (int i = 0; i < stdout.terminalLines; i++) {
      stdout.writeln();
    } // Clear the console
    var speeds = await monitor.getInternetSpeed();
    print("Upload Speed: ${speeds[0]} KB/s");
    print("Download Speed: ${speeds[1]} KB/s");
    await Future.delayed(Duration(milliseconds: 100)); // Delay for a short time
  });
}
